File: android/app/src/main/kotlin/app/myvitals/ui/common/ActivityCategory.kt
package app.myvitals.ui.common

import androidx.compose.ui.graphics.Color
import app.myvitals.ui.neon.NeonMV

/**
 * One palette for "what kind of activity is this", shared by the year
 * calendar, the activity feed rows and anything else that colour-codes by
 * type.
 *
 * Before this existed the same hexes were written out in three places —
 * the calendar's fill `when` block, its hard-coded legend, and the feed
 * row icons — which is how the legend ended up missing Row and Other
 * while the canvas still drew them.
 *
 * The neon values are not invented here: `StrengthHistory.vue:100-111`
 * already maps strength→magenta, yoga→periwinkle, cardio→cyan in the neon
 * skin, so a neon calendar now agrees with the neon strength charts. The
 * classic values are the exact hexes the calendar drew before, so the
 * default shell is unchanged.
 */
enum class ActivityCategory(
    val key: String,
    val label: String,
    val classic: Color,
    val neon: Color,
) {
    STRENGTH("strength", "Strength", Color(0xFFEF4444), NeonMV.Magenta),
    YOGA("yoga", "Yoga", Color(0xFFA78BFA), NeonMV.Periwinkle),
    RIDE("ride", "Ride", Color(0xFF38BDF8), NeonMV.Cyan),
    RUN("run", "Run", Color(0xFF22C55E), NeonMV.Lime),
    WALK("walk", "Walk", Color(0xFFF59E0B), NeonMV.Amber),
    // Teal, deliberately distinct from Cyan — rowing next to a ride must
    // not read as the same category on a 10 px calendar cell.
    ROW("row", "Row", Color(0xFF06B6D4), Color(0xFF2EE6C5)),
    OTHER("other", "Other", Color(0xFF94A3B8), NeonMV.Muted);

    fun color(neon: Boolean): Color = if (neon) this.neon else classic
}

/** Colour for a day with nothing logged. */
fun activityCategoryEmpty(neon: Boolean): Color =
    if (neon) Color(0xFF161A24) else Color(0x141A2332)

/**
 * Classify a raw activity `type` string.
 *
 * The exact matches are the ones the calendar already recognised. The
 * substring fallbacks are new: the catalog carries ~37 distinct type
 * strings across Strava, Fitbit and MyFitnessPal, so variants like
 * `VirtualRide`, `TrailRun` and the MyFitnessPal
 * `walking,_3.0_mph,_mod._pace_...` labels used to fall through to OTHER
 * and render grey.
 *
 * Order matters: "run" is tested before "walk" because several
 * MyFitnessPal labels contain both words.
 */
fun categoryForActivityType(type: String?): ActivityCategory {
    val t = (type ?: "").lowercase()
    return when (t) {
        "ride", "ebikeride", "mountain_biking", "cycling" -> ActivityCategory.RIDE
        "run", "trailrun", "running" -> ActivityCategory.RUN
        "hike", "walk", "walking" -> ActivityCategory.WALK
        "rower", "rowing", "row" -> ActivityCategory.ROW
        else -> when {
            t.isEmpty() -> ActivityCategory.OTHER
            t.contains("kayak") || t.contains("row") || t.contains("paddle") ||
                t.contains("canoe") -> ActivityCategory.ROW
            t.contains("run") || t.contains("jog") -> ActivityCategory.RUN
            t.contains("hike") || t.contains("walk") -> ActivityCategory.WALK
            t.contains("bike") || t.contains("cycl") || t.contains("ride") ->
                ActivityCategory.RIDE
            t.contains("strength") || t.contains("weight") ||
                t.contains("lifting") -> ActivityCategory.STRENGTH
            t.contains("yoga") || t.contains("pilates") ||
                t.contains("mobility") -> ActivityCategory.YOGA
            else -> ActivityCategory.OTHER
        }
    }
}

/**
 * Classify a generated workout by its `split_focus`.
 *
 * Cardio days map to RIDE rather than STRENGTH: the activity feed already
 * draws them with a bike icon (ActivitiesScreen ActivityIcon), so painting
 * them red on the calendar contradicted the row two inches below.
 */
fun categoryForSplitFocus(focus: String?): ActivityCategory =
    when ((focus ?: "").lowercase()) {
        "yoga" -> ActivityCategory.YOGA
        "cardio" -> ActivityCategory.RIDE
        else -> ActivityCategory.STRENGTH
    }

File: frontend/src/utils/activityCategory.ts
/**
 * One palette for "what kind of activity is this" — the web twin of
 * `android/.../ui/common/ActivityCategory.kt`. Both must agree, or the
 * same ride is blue on one surface and cyan on the other.
 *
 * The neon hexes are not new: `StrengthHistory.vue` already maps
 * strength→magenta, yoga→periwinkle, cardio→cyan in the neon skin. The
 * classic hexes are the ones the old Android calendar drew.
 */
export type ActivityCategoryKey =
  | "strength" | "yoga" | "ride" | "run" | "walk" | "row" | "other";

export interface CategoryMeta {
  label: string;
  /** 1-based, used as the heatmap's numeric bucket for the piecewise visualMap. */
  index: number;
  classic: string;
  neon: string;
}

export const CATEGORY_META: Record<ActivityCategoryKey, CategoryMeta> = {
  strength: { label: "Strength", index: 1, classic: "#ef4444", neon: "#ff3ad8" },
  yoga:     { label: "Yoga",     index: 2, classic: "#a78bfa", neon: "#6f7bff" },
  ride:     { label: "Ride",     index: 3, classic: "#38bdf8", neon: "#28e6ff" },
  run:      { label: "Run",      index: 4, classic: "#22c55e", neon: "#5dff3b" },
  walk:     { label: "Walk",     index: 5, classic: "#f59e0b", neon: "#ffb52e" },
  // Teal, deliberately distinct from ride's cyan at calendar-cell size.
  row:      { label: "Row",      index: 6, classic: "#06b6d4", neon: "#2ee6c5" },
  other:    { label: "Other",    index: 7, classic: "#94a3b8", neon: "#9b9bb0" },
};

export const CATEGORY_BY_INDEX: Record<number, ActivityCategoryKey> =
  Object.fromEntries(
    (Object.keys(CATEGORY_META) as ActivityCategoryKey[])
      .map((k) => [CATEGORY_META[k].index, k]),
  ) as Record<number, ActivityCategoryKey>;

/**
 * Classify a raw activity `type`. Exact matches first (these are the ones
 * the old calendar recognised), then substring fallbacks so Strava's
 * VirtualRide / TrailRun and the MyFitnessPal labels stop landing in
 * "other". "run" is tested before "walk" — several MyFitnessPal labels
 * contain both words.
 */
export function categoryForType(
  type: string | null | undefined,
): ActivityCategoryKey {
  const t = (type ?? "").toLowerCase();
  switch (t) {
    case "ride": case "ebikeride": case "mountain_biking": case "cycling":
      return "ride";
    case "run": case "trailrun": case "running":
      return "run";
    case "hike": case "walk": case "walking":
      return "walk";
    case "rower": case "rowing": case "row":
      return "row";
  }
  if (!t) return "other";
  if (/kayak|row|paddle|canoe/.test(t)) return "row";
  if (/run|jog/.test(t)) return "run";
  if (/hike|walk/.test(t)) return "walk";
  if (/bike|cycl|ride/.test(t)) return "ride";
  if (/strength|weight|lifting/.test(t)) return "strength";
  if (/yoga|pilates|mobility/.test(t)) return "yoga";
  return "other";
}

/**
 * Classify a generated workout by split_focus. Cardio maps to "ride", not
 * "strength" — the feed already draws cardio days with a bike icon.
 */
export function categoryForSplitFocus(
  focus: string | null | undefined,
): ActivityCategoryKey {
  const f = (focus ?? "").toLowerCase();
  if (f === "yoga") return "yoga";
  if (f === "cardio") return "ride";
  return "strength";
}

export function categoryColor(k: ActivityCategoryKey, neon: boolean): string {
  const m = CATEGORY_META[k];
  return neon ? m.neon : m.classic;
}
